File: Codes/1706-where-will-the-ball-fall_20250215.java
/*
 * @Author: LetMeFly
 * @Date: 2025-02-15 11:20:57
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-02-15 11:30:31
 */
class Solution {
    private int[][] grid;

    private int drop(int y) {
        for (int x = 0; x < grid.length; x++) {
            int nextY = grid[x][y] + y;
            if (nextY < 0 || nextY >= grid[0].length || grid[x][y] != grid[x][nextY]) {
                return -1;
            }
            y = nextY;
        }
        return y;
    }

    public int[] findBall(int[][] grid) {
        this.grid = grid;
        int[] ans = new int[grid[0].length];
        for (int j = 0; j < grid[0].length; j++) {
            ans[j] = drop(j);
        }
        return ans;
    }
}

File: Codes/1706-where-will-the-ball-fall_20250215.py
'''
Author: LetMeFly
Date: 2025-02-15 11:20:53
LastEditors: LetMeFly.xyz
LastEditTime: 2025-02-15 11:24:22
Description: AC,67.31%,26.92%
'''
from typing import List

class Solution:
    def drop(self, j: int) -> None:
        y = j
        for x in range(len(self.grid)):
            nextY = self.grid[x][y] + y
            if nextY < 0 or nextY >= len(self.grid[0]) or self.grid[x][y] != self.grid[x][nextY]:
                return
            y = nextY
        self.ans[j] = y        

    def findBall(self, grid: List[List[int]]) -> List[int]:
        self.grid = grid
        self.ans = [-1] * len(grid[0])
        for j in range(len(grid[0])):
            self.drop(j)
        return self.ans